test/mooncake.jl

```diff
@@ -0,0 +1,542 @@
+    dA_copy = make_mooncake_tangent(copy(ΔA))
+    dA_inplace = make_mooncake_tangent(copy(ΔA))
+
+    dargs_copy = Δargs isa Tuple ? make_mooncake_fdata.(deepcopy(Δargs)) : make_mooncake_fdata(deepcopy(Δargs))
+
+    dargs_inplace = Δargs isa Tuple ? make_mooncake_fdata.(deepcopy(Δargs)) : make_mooncake_fdata(deepcopy(Δargs))
+
+    copy_out, copy_pb!! = isnothing(alg) ? rrule(Mooncake.CoDual(f_c, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_copy), Mooncake.CoDual(args, dargs_copy)) : rrule(Mooncake.CoDual(f_c, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_copy), Mooncake.CoDual(args, dargs_copy), Mooncake.CoDual(alg, Mooncake.NoFData()))
+    copy_pb!!(rdata)
+
+    # not every f! has a handwritten rrule!!
+    inplace_sig = isnothing(alg) ? Tuple{typeof(f!), typeof(A), typeof(args)} : Tuple{typeof(f!), typeof(A), typeof(args), typeof(alg)}
+
+    has_handwritten_rule = hasmethod(Mooncake.rrule!!, inplace_sig)
+    if has_handwritten_rule
+        inplace_out, inplace_pb!! = isnothing(alg) ? Mooncake.rrule!!(Mooncake.CoDual(f!, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_inplace), Mooncake.CoDual(args, dargs_inplace)) : Mooncake.rrule!!(Mooncake.CoDual(f!, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_inplace), Mooncake.CoDual(args, dargs_inplace), Mooncake.CoDual(alg, Mooncake.NoFData()))
+    else
+        rvs_interp = Mooncake.get_interpreter(Mooncake.ReverseMode)
+        inplace_rrule = Mooncake.build_rrule(rvs_interp, inplace_sig)
+        inplace_out, inplace_pb!! = isnothing(alg) ? inplace_rrule(Mooncake.CoDual(f!, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_inplace), Mooncake.CoDual(args, dargs_inplace)) : inplace_rrule(Mooncake.CoDual(f!, Mooncake.NoFData()), Mooncake.CoDual(Ac, dA_inplace), Mooncake.CoDual(args, dargs_inplace), Mooncake.CoDual(alg, Mooncake.NoFData()))
+    end
+    inplace_pb!!(rdata)
+
+    dA_inplace_ = Mooncake.arrayify(A, dA_inplace)[2]
+    dA_copy_ = Mooncake.arrayify(A, dA_copy)[2]
+    @test dA_inplace_ ≈ dA_copy_
+    return
+end
+
+@timedtestset "QR AD Rules with eltype $T" for T in ETs
+    rng = StableRNG(12345)
+    m = 19
+    @testset "size ($m, $n)" for n in (17, m, 23)
+        atol = rtol = m * n * precision(T)
+        A = randn(rng, T, m, n)
+        minmn = min(m, n)
+        @testset for alg in (
+                LAPACK_HouseholderQR(),
+                LAPACK_HouseholderQR(; positive = true),
+            )
+            @testset "qr_compact" begin
+                QR = qr_compact(A, alg)
+                Q = randn(rng, T, m, minmn)
+                R = randn(rng, T, minmn, n)
+                Mooncake.TestUtils.test_rule(rng, qr_compact, A, alg; mode = Mooncake.ReverseMode, is_primitive = false, atol = atol, rtol = rtol)
+                qr_c = (A, QR, alg) -> qr_compact!(MatrixAlgebraKit.copy_input(qr_compact, A), QR, alg)
+                Mooncake.TestUtils.test_rule(rng, qr_c, A, (Q, R), alg; mode = Mooncake.ReverseMode, is_primitive = false, atol = atol, rtol = rtol)
```

Comment:
What is the logic for this extra test, which you don't seem to have for the other functions?

Reply:
Ah, it's a proof of concept relic from me developing the `test_pullbacks_match`